### DISCUSSION TITLE
AnalogInput signal looks different from input signal from BNC

### DISCUSSION THREAD

**Discussion by sarahapark** [bonsai-rx/bonsai#877] [Q&A] [answered]
Hello!
I am sending input signal to AnalogInput via a BNC with the signal being recorded by Bonsai and ThorSync. (I have the signal recorded by two methods so as to have a Bonsai backup when ThorSync sometimes stops recording in the middle of an experiment. Also, Bonsai is synced to video recordings for behavior.)

When I overlaid the same signal from Bonsai (figure: red trace) and ThorSync (figure: blue trace), they are different. What could be the problem? How can I fix the problem?

![TSandBonsaiOverlay_VRY](https://user-images.githubusercontent.com/51828933/164023062-3b79521f-b2cf-41f4-bc5a-56d6ee3f3267.jpg)

Thank you!
Sarah

PS - I tried to attach the bonsai code but it is not a supported file type. How do I attach the Bonsai file in github?

**Comment by bruno-f-cruz** [comment]
You can attach a .bonsai by zipping it first. 

Regarding the difference in values, what are you plotting in the Y and X axis exactly? I am wondering if it is a matter of both signal and time resolution of the two ADCs perhaps?

**Comment by sarahapark** [comment, marked as answer]
Here's the Bonsai file.

I'm tracking VR position on the y axis and time on the x axis.

[NI6001v4CameraAndNIDAQ.zip](https://github.com/bonsai-rx/bonsai/files/8514343/NI6001v4CameraAndNIDAQ.zip)
.

**Reply by bruno-f-cruz** [reply]
Are the resolution (probably bitdepth?) and sampling rate the same in the two graphs?

**Reply by sarahapark** [reply]
Thorsync is 16-bit. (From ThorLabs: "Analog samples have a 16 bit resolution across a +/-10V range.")
The sampling rate is 1kHz for both ThorSync and Bonsai.

**Reply by bruno-f-cruz** [reply]
Looking at the graph it seems like you are rounding the values coming out of the ADC to the nearest integer (I see a total of 10 evenly spaced values). I dont  have a nidaq with me right now, but i suspect the output is a float already in S.I. units (V). If you can confirm this, try removing the 'ConvertScale' after the 'AnalogInput' that is effectively rounding the incoming values by converting them to S16. When opening the binary file later, do not forget to adjust your data format accordingly.

**Reply by sarahapark** [reply]
What are ways I can check that the output is a float?

**Reply by bruno-f-cruz** [reply]
You can right click on the AnalogOutout node.

**Reply by sarahapark** [reply]
Hmmm...I don't have an enabled **AnalogOutput** node. Did you mean the **AnalogInput** node or another node? 

When I comment out **ConvertScale**, I get the error message that input arrays must have the same element type.

**Reply by bruno-f-cruz** [reply]
I meant 'AnalogInput', my bad. The error you are getting is due to the 'Concat' node in your workflow. To concatenate tha matricies they should have the same datatype. For now, let's stick with your initial issue. You can simply create a parallel branch that saves your incoming analog data as a floating point. Send the output of 'AnalogRead' to a 'MatrixWritter' and assert if you have the same problem as before or if the data matches.

**Reply by sarahapark** [reply]
Hi Bruno,
First, **AnalogInput** is reading floating point (F64).
Second, I tried to create a parallel **AnalogInput** to **MatrixWriter** (see attached code), but the second **MatrixWriter** doesn't save anything (0 bytes saved file). The parallel **AnalogInput** and **MatrixWriter** nodes are not receiving any data -- the **TextObjectVisualizer** remains blank.
Third, I am timestamping and time syncing the video frames with the analog data, so any changes I make will need to maintain this time sync and timestamping for analysis later on.

What do you suggest I try next?
[NI6001v5CameraAndNIDAQ - draft.zip](https://github.com/bonsai-rx/bonsai/files/8555503/NI6001v5CameraAndNIDAQ.-.draft.zip)

**Reply by bruno-f-cruz** [reply]
Can you try to save it using a single branch. Something like this:
[NI6001v5CameraAndNIDAQ - draft.zip](https://github.com/bonsai-rx/bonsai/files/8555952/NI6001v5CameraAndNIDAQ.-.draft.zip)
And try to check the differences inthe output?

**Reply by sarahapark** [reply]
Here's the output using the Bonsai code you suggested. The Bonsai data still looks different from the ThorSync data. I am wondering if my MatLab code is wrong, so I'm including the MatLab code and the plotted data.

First, here's the plot of the ThorSync data:
![image](https://user-images.githubusercontent.com/51828933/165346784-a0ee6af8-71c9-4e97-b0de-034fb2602b26.png)

Second, here are the MatLab code and plot of the Bonsai data saved with **MatrixWriter** as S16 and read as "int16" in MatLab:
![image](https://user-images.githubusercontent.com/51828933/165347786-3190cdc2-e8db-4c7c-92b1-bcff482b7cf9.png)
![image](https://user-images.githubusercontent.com/51828933/165347812-b1fe0a63-9d59-4ab1-9c41-c0ab1f46f092.png)

Third, here are the MatLab code and plot of the Bonsai data saved with **MatrixWriter** as floating point and read as "double" in Matlab:
![image](https://user-images.githubusercontent.com/51828933/165347103-6b30881b-ca40-4d4e-bbd8-cfa0a9fb31de.png)
![image](https://user-images.githubusercontent.com/51828933/165347145-dff1a0b0-2b17-49f3-a6f2-346890a75d5a.png)

Fourth, here are the MatLab code and plot of the Bonsai data saved with **MatrixWriter** as floating point and read as "float" in Matlab:
![image](https://user-images.githubusercontent.com/51828933/165348275-d8fa9c7f-d9ed-4b9c-97c2-502c38d6a138.png)
![image](https://user-images.githubusercontent.com/51828933/165348298-5ead51a7-f89a-46c5-9cd4-8fb96abfe408.png)

**Reply by bruno-f-cruz** [reply]
`binChannels` should be 8, not 9. In the workflow I sent you you are saving all Analog Channels (which I am assuming in your case are 8?) but you are not concatenating the digital ones. Can you make that small change and try again? 

Sorry for the confusion

**Reply by sarahapark** [reply]
Bingo! We got it to work! Thank you so much for your help!

![image](https://user-images.githubusercontent.com/51828933/165559143-bdbac1a2-d598-41e7-9dc5-051ee9aba749.png)
